Geometry/HGCalMapping/test/alpaka/HGCalMappingESSourceTester.cc

```diff
@@ -0,0 +1,446 @@
+      : cellIndexTkn_(esConsumes<HGCalMappingCellIndexer, HGCalElectronicsMappingRcd>()),
+        cellTkn_(esConsumes(edm::ESInputTag(""))),
+        moduleIndexTkn_(esConsumes<HGCalMappingModuleIndexer, HGCalElectronicsMappingRcd>()),
+        moduleTkn_(esConsumes(edm::ESInputTag(""))),
+        testCollToken_{produces()} {}
+
+  //
+  void HGCalMappingESSourceTester::produce(device::Event& iEvent, device::EventSetup const& iSetup) {
+    //put a dummy collection to the event
+    portabletest::TestDeviceCollection testColl{0, iEvent.queue()};
+    iEvent.emplace(testCollToken_, std::move(testColl));
+
+    // if the cfg didn't change there's nothing else to do
+    if (!cfgWatcher_.check(iSetup))
+      return;
+
+    //get cell indexers and SoA
+    auto cellIdx = iSetup.getData(cellIndexTkn_);
+    auto const& cells = iSetup.getData(cellTkn_);
+    edm::LogInfo("HGCalMappingIndexESSourceTester") << "Cell dense indexers and associated SoA retrieved for HGCAL";
+    int nmodtypes = cellIdx.typeCodeIndexer_.size();
+    edm::LogInfo("HGCalMappingIndexESSourceTester")
+        << "[Module cell indexer] has " << nmodtypes << " module types" << std::endl;
+
+    //printout and test the indexer contents for cells
+    edm::LogInfo("HGCalMappingIndexESSourceTester").log([&](auto& log) {
+      uint32_t totOffset(0);
+      for (size_t idx = 0; idx < cellIdx.di_.size(); idx++) {
+        //check typecode exists
+        auto typecode = cellIdx.getTypecodeFromEnum(idx);
+        assert(cellIdx.typeCodeIndexer_.count(typecode) == 1);
+
+        //check that the current offset is consistent with the increment from cells from the previous module
+        if (idx > 0) {
+          uint32_t nch_prev = cellIdx.maxErx_[idx - 1] * cellIdx.maxChPerErx_;
+          uint32_t delta_offset = cellIdx.offsets_[idx] - cellIdx.offsets_[idx - 1];
+          assert(delta_offset == nch_prev);
+        }
+
+        //assert offset is consistent with the accumulation
+        auto off = cellIdx.offsets_[idx];
+        assert(off == totOffset);
+
+        totOffset += cellIdx.maxErx_[idx] * cellIdx.maxChPerErx_;
+
+        //print
+        log << "\t [" << typecode << "] has "
+            << " index(internal)=" << idx << " #eRx = " << cellIdx.maxErx_[idx]
+            << " #cells = " << cellIdx.di_[idx].getMaxIndex() << " offset @ " << cellIdx.offsets_[idx] << "\n";
+      }
+
+      assert(totOffset == cellIdx.maxDenseIndex());
+      log << "SoA size for module cell mapping will be " << totOffset << "\n";
+    });
+
+    //printout and test module cells SoA contents
+    uint32_t ncells = cells.view().metadata().size();
+    assert(ncells == cellIdx.maxDenseIndex());  //check for consistent size
+    LogDebug("HGCalMappingIndexESSourceTester").log([&](auto& log) {
+      log << "Module cell mapping contents \n";
+      for (uint32_t i = 0; i < ncells; i++) {
+        log << "\tidx = " << i << ", "
+            << "isHD = " << cells.view()[i].isHD() << ", "
```

Comment:
ok i didn't realize LogDebug would not work on non-CPU. What's the guideline in this case?

Reply:
So far `printf` is the only way of printing messages from a non-CPU device.

Reply:
To be more precise, all accesses to the device-side data must be done in a kernel (or in a device-side function called from a kernel). In this specific case, you can pass the view around (i.e. return value of `cells.view()`), and _maybe_ `cells.view()[i]` is still fine (because _maybe_ it  does only pointer arithmetic on the device-side data), but access to a value in `cells.view()[i].isHD()` will certainly result in invalid memory access on non-CPU backends.

Reply:
OK - on this one, and only if there are no additional comments on the rest, can we leave it as is for the moment? 
The reasoning is the following:
- it's a tester plugin to inspect the contents of the records, it does it job on CPU
- we may need more time to do the remaining changes to run with a GPU [1]
- we would like to have the rest of the code already available in release to trigger a series of parallel efforts on the DIGI-RAW-DIGI chain

[1] My personal bet is that after the Patatrack Hackhathon (March) we'll have it updated

Reply:
> it's a tester plugin to inspect the contents of the records, it does it job on CPU

Can it simply be a non-portable `edm::stream::EDAnalyzer` then ?

Reply:
maybe it's enough as analyzer...